### PR TITLE
Restart bugfix

### PR DESCRIPTION
**Pull request by nmsriram**
## PR Summary
This merge fixes a parallel restart bug in the restart path.  Parallel restarts with same (or different) number of processors now work for the advection example.  

Also fixes a restart bug with reflective boundary conditions with an ugly hack in `mesh/mesh_refinement.cpp` to apply reflective boundary conditions.  It is expected that @Yurlungur or @jonahm-LANL will fix this when doing the boundary condition refactor.  As a result this should get merged in before the boundary condition fix.

We can now restart with a number of processors different from the original run, and even restart serial from parallel (or vice versa) and still get identical answers.

I believe #61 is fixed with this PR.

Known issue: 

## PR Checklist

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md

### CODE REVIEW

File: src/mesh/mesh.cpp
Comment (JoshuaSBrown) [inline src/mesh/mesh.cpp]: Isn't this redundant with the initialization above?

File: src/mesh/mesh.hpp
Comment (JoshuaSBrown) [inline src/mesh/mesh.hpp]: This should at least be made const.

File: src/mesh/mesh_refinement.hpp
Comment (JoshuaSBrown) [inline src/mesh/mesh_refinement.hpp]: This can be made const noexcept.

File: src/mesh/mesh_refinement.hpp
Comment (JoshuaSBrown) [inline src/mesh/mesh_refinement.hpp]: **a** should be const reference.

File: src/outputs/restart.hpp
Comment (JoshuaSBrown) [inline src/outputs/restart.hpp]: Good catch! Oh nevermind, I guess this didn't actually really change anything...

File: src/parthenon_manager.cpp
Comment (JoshuaSBrown) [inline src/parthenon_manager.cpp]: Remove comment?

File: src/mesh/mesh.cpp
Comment (JoshuaSBrown) [inline src/mesh/mesh.cpp]: It looks like mrI can be made const.

**Review by Yurlungur — APPROVED** [review]
Thanks for the fix. I don't love the function name `GetNbs` as `nb*` is a cryptic name. But since that's what they're called internally, I think it's fine.

**Review by JoshuaSBrown — COMMENTED** [review]
I'm still getting errors when running the tests, which tests is this supposed to fix?

```Start 27: regression_test:advection_performance
27/36 Test #27: regression_test:advection_performance .............................***Failed  336.32 sec
      Start 28: regression_mpi_test:advection_performance
28/36 Test #28: regression_mpi_test:advection_performance .........................***Failed  412.27 sec
      Start 29: regression_test:restart
29/36 Test #29: regression_test:restart ...........................................   Passed   69.50 sec
      Start 30: regression_mpi_test:restart
30/36 Test #30: regression_mpi_test:restart .......................................***Failed    1.38 sec
```

**Comment by JoshuaSBrown** [comment]
It looks like the test will run but something is still throwing an error:

```
/projects/opt/ppc64le/p9/openmpi/4.0.2-gcc_7.4.0/bin/mpiexec -n 2 '/home/joshuasbrown/Software/parthenon_x86_64/parthenon/build/example/advection/advection-example' -i '/home/joshuasbrown/Software/parthenon_x86_64/parthenon/tst/regression/test_suites/restart/parthinput.restart' 'parthenon/job/problem_id=gold'
--------------------------------------------------------------------------
WARNING: There was an error initializing an OpenFabrics device.

  Local host:   cn2020
  Local device: mlx5_1
--------------------------------------------------------------------------
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false
### Warning in Mesh::Initialize
The number of MeshBlocks increased more than twice during initialization.
More computing power than you expected may be required.
### Warning in Mesh::Initialize
The number of MeshBlocks increased more than twice during initialization.
More computing power than you expected may be required.

Setup complete, executing driver...

cycle=0 time=0.0000000000000000e+00 dt=1.7578125000000000e-03 zone-cycles/wsec = 0.00e+00
cycle=1 time=1.7578125000000000e-03 dt=1.7578125000000000e-03 zone-cycles/wsec = 1.85e+04
cycle=2 time=3.5156250000000001e-03 dt=1.7578125000000000e-03 zone-cycles/wsec = 1.36e+05
cycle=3 time=5.2734375000000003e-03 dt=1.7578125000000000e-03 zone-cycles/wsec = 6.90e+04
[cn2020:30009] 1 more process has sent help message help-mpi-btl-openib.txt / error in device init
[cn2020:30009] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
```

File: src/mesh/mesh.hpp
Comment (nmsriram) [reply src/mesh/mesh.hpp]: const wouldn't do anything since the method returns a scalar `int`, right?

Comment (Yurlungur) [reply src/mesh/mesh.hpp]: I think @JoshuaSBrown means `int GetNBs() const { ... }`, which would enforce that the method does not change the state of the `Mesh` object.

Comment (JoshuaSBrown) [reply src/mesh/mesh.hpp]: You are right, it would just show intent, but it will also allow you to call the method from within another const method. @Yurlungur thanks for clarifying.

**Review by JoshuaSBrown — APPROVED** [review]
Other than my optional improvements this is good. The error I was running into is located in the python regression testing framework so this has my green light.

**Comment by Yurlungur** [comment]
@nmsriram is this ready for merge?

File: src/mesh/mesh.cpp
Comment (pgrete) [reply src/mesh/mesh.cpp]: I think so, too.

File: src/mesh/mesh_refinement.hpp
Comment (pgrete) [inline src/mesh/mesh_refinement.hpp]: Do we need a `vector` here? Could this just be an `array`?

Comment (pgrete) [inline src/mesh/mesh_refinement.hpp]: Given that `int32_t` is used above to read, I suggest to also be more specific here.

File: src/outputs/restart.hpp
Comment (pgrete) [reply src/outputs/restart.hpp]: Is this `ifdef` behavior actually what we want? (and/or is this a leftover from Athena++?)
I'm asking as having the code fail seems like a preferable solution rather than returning an empty object (due to lack of implementation or similar).

File: src/mesh/mesh.cpp
Comment (nmsriram) [reply src/mesh/mesh.cpp]: This was me clutching at straws trying to figure out why MPI on more than original number of nodes still is incorrect and why the LANL physics code has the ghost cell problem.

File: src/mesh/mesh_refinement.hpp
Comment (nmsriram) [reply src/mesh/mesh_refinement.hpp]: Will do.

**Comment by nmsriram** [comment]
@pgrete the reason to return an empty array is that we may change the physics on restart and I didn't want the request for those physics to kill the code.  I could go either way on this, perhaps with a query function.

**Comment by pgrete** [comment]
@nmsriram there are a couple of comments left. It'd be great if they could be addressed (or replied to) before we go ahead and merge this.

Also (unrelated to this PR): I just noticed that quite a bit of the restart functionality has been implemented using macros.
What's the motivation behind those macros and could we potentially replace those (not in this PR) with standard functions?

File: src/mesh/mesh.cpp
Comment (nmsriram) [reply src/mesh/mesh.cpp]: Removed.

File: src/mesh/mesh.hpp
Comment (nmsriram) [reply src/mesh/mesh.hpp]: This has been removed.

File: src/parthenon_manager.cpp
Comment (nmsriram) [reply src/parthenon_manager.cpp]: GetNbs() was removed and line uncommented.

File: src/mesh/mesh.cpp
Comment (nmsriram) [reply src/mesh/mesh.cpp]: removed.

File: src/mesh/mesh_refinement.hpp
Comment (nmsriram) [reply src/mesh/mesh_refinement.hpp]: This has been removed.

**Comment by nmsriram** [comment]
@pgrete the reason they are macros is because I couldn't figure out how to do it different. About a week ago @AndrewGaspar commented essentially the same thing and showed a std::function way.  I'm a bit hesitant to change this code at this point and will do it in a subsequent PR.

File: src/mesh/mesh_refinement.hpp
Comment (nmsriram) [reply src/mesh/mesh_refinement.hpp]: @JoshuaSBrown the restart test passes now in both serial and parallel, and I changed 'gold' to 'silver'.

Comment (nmsriram) [reply src/mesh/mesh_refinement.hpp]: Also getInternals has been removed.

File: src/interface/variable_pack.hpp
Comment (Yurlungur) [inline src/interface/variable_pack.hpp:157]: We should perhaps formalize this throughout the code and make a habit of initializing to [signal NaNs](https://en.cppreference.com/w/cpp/types/numeric_limits/signaling_NaN). That's for another time, though.

File: src/mesh/mesh_refinement.cpp
Comment (Yurlungur) [inline src/mesh/mesh_refinement.cpp:41]:
```suggestion
// TODO(JMM) Identify an idiomatic way of applying reflective
```

File: src/mesh/mesh_refinement.cpp
Comment (Yurlungur) [inline src/mesh/mesh_refinement.cpp:86]: This actually needs to be done for reflecting, outflow, and "user," not just reflecting. But I'll deal with it in my MR.

File: tst/regression/test_suites/restart/restart.py
Comment (Yurlungur) [inline tst/regression/test_suites/restart/restart.py:56]: This is a good and important fix. But it may conflict with the same fix in #330

File: src/parthenon_manager.cpp
Comment (Yurlungur) [inline src/parthenon_manager.cpp:237]: is this a debug statement that should be removed?

**Review by Yurlungur — APPROVED** [review]
nice work. this one was a doozy.

File: src/mesh/mesh_refinement.cpp
Comment (nmsriram) [reply src/mesh/mesh_refinement.cpp:86]: Thanks!

File: tst/regression/test_suites/restart/restart.py
Comment (nmsriram) [reply tst/regression/test_suites/restart/restart.py:56]: #330 makes a few more deeper changes to the structure and comparisons, and will override this one.  Without this fix in place I couldn't test which is why I made it here as well.

File: src/parthenon_manager.cpp
Comment (nmsriram) [reply src/parthenon_manager.cpp:237]: I put it in there intentionally.  I wanted to give out a status reading that variables are read in from restart file so that the outputs from a direct run and a restart run are evident, and also so that somebody can check to see what variables came in from the restart file.

Comment (Yurlungur) [reply src/parthenon_manager.cpp:237]: Ah, okay. Got it.

File: src/parthenon_manager.cpp
Comment (pgrete) [inline src/parthenon_manager.cpp:195]:
```suggestion
```

**Review by pgrete — APPROVED** [review]

File: src/interface/variable_pack.hpp
Comment (JoshuaSBrown) [reply src/interface/variable_pack.hpp:157]: Should we create an issue?

File: src/mesh/mesh_refinement.cpp
Comment (JoshuaSBrown) [inline src/mesh/mesh_refinement.cpp:178]: Why do we only have a function like this for the 2d case?

File: src/outputs/parthenon_hdf5.cpp
Comment (JoshuaSBrown) [inline src/outputs/parthenon_hdf5.cpp:202]: Nice!

File: src/outputs/restart.cpp
Comment (JoshuaSBrown) [inline src/outputs/restart.cpp:246]: These comments are confusing to me, what does the SS mean, this kind of looks like you are saying nx1 is the same as mb.block_size.nx1. If so why are you using mb.block_size.nx1 to define bsize and not the more concise nx1 nx2 and nx3 variables? 

Ok, I see, nx1 etc will change depending on theDomain.

File: src/outputs/restart.hpp
Comment (JoshuaSBrown) [inline src/outputs/restart.hpp:183]: Why int instead of bool?

**Comment by JoshuaSBrown** [comment]
```
    Start  1: unit_tests:Just check everything
 1/37 Test  #1: unit_tests:Just check everything ..................................   Passed    0.27 sec
      Start  2: unit_tests:Task Object Lifecycle
 2/37 Test  #2: unit_tests:Task Object Lifecycle ..................................   Passed    0.27 sec
      Start  3: unit_tests:Can create a vector-valued face-variable
 3/37 Test  #3: unit_tests:Can create a vector-valued face-variable ...............   Passed    0.25 sec
      Start  4: unit_tests:Add and Get is called
 4/37 Test  #4: unit_tests:Add and Get is called ..................................   Passed    0.26 sec
      Start  5: unit_tests:reset is called
 5/37 Test  #5: unit_tests:reset is called ........................................   Passed    0.26 sec
      Start  6: unit_tests:when hasKey is called
 6/37 Test  #6: unit_tests:when hasKey is called ..................................   Passed    0.29 sec
      Start  7: unit_tests:Physical constants
 7/37 Test  #7: unit_tests:Physical constants .....................................   Passed    0.29 sec
      Start  8: unit_tests:Checking IndexShape indices
 8/37 Test  #8: unit_tests:Checking IndexShape indices ............................   Passed    0.25 sec
      Start  9: unit_tests:Checking IndexShape cell counts
 9/37 Test  #9: unit_tests:Checking IndexShape cell counts ........................   Passed    0.24 sec
      Start 10: unit_tests:par_for loops
10/37 Test #10: unit_tests:par_for loops ..........................................   Passed    0.61 sec
      Start 11: unit_tests:nested par_for loops
11/37 Test #11: unit_tests:nested par_for loops ...................................   Passed    0.29 sec
      Start 12: unit_tests:Overlapping SpaceInstances
12/37 Test #12: unit_tests:Overlapping SpaceInstances .............................   Passed    0.34 sec
      Start 13: unit_tests:Built-in flags are registered
13/37 Test #13: unit_tests:Built-in flags are registered ..........................   Passed    0.34 sec
      Start 14: unit_tests:A Metadata flag is allocated
14/37 Test #14: unit_tests:A Metadata flag is allocated ...........................   Passed    0.26 sec
      Start 15: unit_tests:A Metadata struct is created
15/37 Test #15: unit_tests:A Metadata struct is created ...........................   Passed    0.25 sec
      Start 16: unit_tests:ParArrayND
16/37 Test #16: unit_tests:ParArrayND .............................................   Passed    0.31 sec
      Start 17: unit_tests:ParArrayND with LayoutLeft
17/37 Test #17: unit_tests:ParArrayND with LayoutLeft .............................   Passed    0.35 sec
      Start 18: unit_tests:Time simple stencil operations
18/37 Test #18: unit_tests:Time simple stencil operations .........................   Passed    0.27 sec
      Start 19: unit_tests:Check registry pressure
19/37 Test #19: unit_tests:Check registry pressure ................................   Passed    0.99 sec
      Start 20: unit_tests:Check many arrays
20/37 Test #20: unit_tests:Check many arrays ......................................   Passed    0.26 sec
      Start 21: unit_tests:Swarm memory management
21/37 Test #21: unit_tests:Swarm memory management ................................   Passed    0.31 sec
      Start 22: unit_tests:Can pull variables from containers based on Metadata
22/37 Test #22: unit_tests:Can pull variables from containers based on Metadata ...   Passed    0.67 sec
      Start 23: unit_tests:Test required/desired checking from inputs
23/37 Test #23: unit_tests:Test required/desired checking from inputs .............   Passed    0.26 sec
      Start 24: unit_tests:Parthenon Error Checking
24/37 Test #24: unit_tests:Parthenon Error Checking ...............................   Passed    0.27 sec
      Start 25: unit_tests:Check that integer division, rounded up, works
25/37 Test #25: unit_tests:Check that integer division, rounded up, works .........   Passed    0.27 sec
      Start 26: unit_tests:Check that partitioning a container works
26/37 Test #26: unit_tests:Check that partitioning a container works ..............   Passed    0.26 sec
      Start 27: performance_tests:Catch2 Container Iterator Performance
27/37 Test #27: performance_tests:Catch2 Container Iterator Performance ...........   Passed   11.80 sec
      Start 28: regression_test:advection_performance
28/37 Test #28: regression_test:advection_performance .............................   Passed  334.83 sec
      Start 29: regression_mpi_test:advection_performance

29/37 Test #29: regression_mpi_test:advection_performance .........................   Passed  1095.79 sec
      Start 30: regression_test:restart
30/37 Test #30: regression_test:restart ...........................................   Passed   69.44 sec
      Start 31: regression_mpi_test:restart
31/37 Test #31: regression_mpi_test:restart .......................................   Passed  112.76 sec
      Start 32: regression_test:calculate_pi
32/37 Test #32: regression_test:calculate_pi ......................................   Passed    2.77 sec
      Start 33: regression_mpi_test:calculate_pi
33/37 Test #33: regression_mpi_test:calculate_pi ..................................   Passed    2.92 sec
      Start 34: regression_test:advection_convergence
34/37 Test #34: regression_test:advection_convergence .............................   Passed  376.76 sec
      Start 35: regression_mpi_test:advection_convergence
35/37 Test #35: regression_mpi_test:advection_convergence .........................   Passed  682.34 sec
      Start 36: regression_test:output_hdf5
36/37 Test #36: regression_test:output_hdf5 .......................................   Passed  241.76 sec
      Start 37: regression_mpi_test:output_hdf5
37/37 Test #37: regression_mpi_test:output_hdf5 ...................................   Passed  368.85 sec

```
Congrats! It works.

File: src/mesh/mesh_refinement.cpp
Comment (Yurlungur) [reply src/mesh/mesh_refinement.cpp:178]: good catch. I think this should be for all 3 cases, @nmsriram

File: src/interface/variable_pack.hpp
Comment (Yurlungur) [reply src/interface/variable_pack.hpp:157]: yes good idea

**Review by Yurlungur — APPROVED** [review]
not sure why my approval didn't stick. I approve.

File: src/outputs/restart.hpp
Comment (nmsriram) [reply src/outputs/restart.hpp:183]: Because it is read in from the HDF file and this requires fewer lines of code

File: src/outputs/restart.cpp
Comment (nmsriram) [reply src/outputs/restart.cpp:246]: Sorry, SS are my initials and I use // SS or //SS where I think I may have screwed up and may need to go back.

File: src/mesh/mesh_refinement.cpp
Comment (nmsriram) [reply src/mesh/mesh_refinement.cpp:178]: Oh brother. yes.  of course.  @JoshuaSBrown / @Yurlungur  Should I put in another PR to fix this, or should I just depend on @Yurlungur to fix it?

Comment (JoshuaSBrown) [reply src/mesh/mesh_refinement.cpp:178]: Whoever, gets to it first I suppose.

**Pull request merged**
